components/raftstore/src/store/fsm/peer.rs

```diff
@@ -983,6 +984,37 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
             self.fsm.has_ready = true;
         }
 
+        // If lease expired, we will send a noop read index to renew lease.
+        if self.fsm.peer.is_leader() {
+            let current_time = match self.ctx.current_time {
+                None => {
+                    self.ctx.current_time = Some(monotonic_raw_now());
+                    self.ctx.current_time.unwrap()
+                }
+                Some(t) => t,
+            };
+            let next_tick = current_time + self.ctx.cfg.raft_base_tick_interval();
+            // We need to propose a read index request if current lease can't cover till next tick
+            let need_propose = match self.fsm.peer.leader_lease.inspect(Some(next_tick)) {
+                LeaseState::Expired => {
+                    let max_lease = self.ctx.cfg.raft_store_max_leader_lease();
+                    self.fsm.peer.pending_reads.back_mut().map_or(true, |read| {
+                        // If there are read index whose lease can cover till next tick
+                        // then we don't need to propose a new one
+                        read.renew_lease_time + max_lease < next_tick
+                    })
+                }
+                _ => false,
+            };
+            if need_propose {
+                let (id, dropped) = self.fsm.peer.propose_read_index();
```

Comment:
Instead of read index, can we use raft heartbeat to ensure leadership?

Reply:
Does it break hibernate region?

Reply:
Do you mean by utilizing `check_quorum`? It may possible that the leader's FSM run slower than the follower, and follower `tick` more times than the leader in the same period of physical time.

Reply:
> Do you mean by utilizing check_quorum? It may possible that the leader's FSM run slower than the follower, and follower tick more times than the leader in the same period of physical time.

I see.

What about hibernate region?

Reply:
It won't break hibernate region, the frequency to renew lease in raft tick is long enough to prevent breaking hibernate state.